File: heap/deepDive/recongnize_string.java
// import static org.junit.jupiter.api.Assertions.assertEquals;

// import org.junit.jupiter.api.Test;

/**
Given a string s, rearrange the characters of s so that any two adjacent characters are not the same.

Return any possible rearrangement of s or return "" if not possible.



Example 1:

Input: s = "aab"
Output: "aba"

 1 count freq
 {'a':2, b:1}
Example 2:

Input: s = "aaab"  ab  a   a
Output: ""

1 count freq
 {'a':2, b:1}
First iteration
temp = 'a'  remove a and b from heap and we update ma


1- Count the frequency of each char
2- User our heap to sort the char by freq
3- as long as the heap > 1 
   we remove the 1st and 2st most freq char
   add to temp
   decrease the value in the map


   







 

*/
import java.util.*;

class Main {
  public static void main(String[] args) {
    System.out.println(reorg("aaab"));
  }

  public static String reorg(String s){

  HashMap<Character , Integer> map = new HashMap<>();
    for(char c : s.toCharArray())
      map.put(c, map.getOrDefault(c,0)+1);

    PriorityQueue<Character> qq = new PriorityQueue((n1, n2)-> map.get(n2) - map.get(n1));
    qq.addAll(map.keySet());

    StringBuilder sb = new StringBuilder();

    while(qq.size()>1){
      char c1 = qq.remove();
      char c2 = qq.remove();
      sb.append(c1);
      sb.append(c2);

      map.put(c1, map.get(c1)-1);
      map.put(c2, map.get(c2)-1);

      if(map.get(c1)> 0)
        qq.add(c1);
      if(map.get(c2)> 0)
        qq.add(c2);
    }

    if (!qq.isEmpty()){
      char cn = qq.remove();
      if (map.get(cn)>1)
        return "";
      sb.append(cn);
    }
    return sb.toString();
  }
}

File: heap/deepDive/recongnize_string.py
""" 
Given a string s, rearrange the characters of s so that any two adjacent characters are not the same.

Return any possible rearrangement of s or return "" if not possible.

 

Example 1:

Input: s = "aab"
Output: "aba"

Example 2:

Input: s = "aaab"
Output: ""

"""
import heapq


class Solution:
    def reorganizeString(self, s: str) -> str:
        # Step 1: Count the frequency of each character in the input string
        freq_map = {}
        for c in s:
            freq_map[c] = freq_map.get(c, 0) + 1

        # Step 2: Create a max heap based on character frequencies
        max_heap = [(-freq, char) for char, freq in freq_map.items()]
        heapq.heapify(max_heap)

        # Step 3: Rearrange characters to avoid adjacent duplicates
        result = []
        while len(max_heap) > 1:
            freq1, char1 = heapq.heappop(max_heap)
            freq2, char2 = heapq.heappop(max_heap)

            # Append characters to the result list
            result.extend([char1, char2])

            # Update frequency in the map
            freq_map[char1] -= 1
            freq_map[char2] -= 1

            # Push characters back onto the heap if their frequency is still greater than 0
            if freq_map[char1] > 0:
                heapq.heappush(max_heap, (-freq_map[char1], char1))
            if freq_map[char2] > 0:
                heapq.heappush(max_heap, (-freq_map[char2], char2))

        # Check if there is a single character left in the heap
        if max_heap:
            freq, last_char = heapq.heappop(max_heap)
            if freq < -1:
                return ""
            result.append(last_char)

        # Step 4: Return the rearranged string
        return ''.join(result)
